File: sdks/node/src/predicates.ts
/** What a predicate sees when a job is being enqueued. */
export interface PredicateContext {
  readonly taskName: string;
  /** The positional args (after any `onEnqueue` rewrites). */
  readonly args: readonly unknown[];
}

/**
 * A gate evaluated at enqueue time. Returning `false` rejects the submission
 * with a {@link PredicateRejectedError}. Register one with {@link Queue.gate}.
 */
export type Predicate = (ctx: PredicateContext) => boolean;

/** A predicate that passes only when every `predicates` member passes. */
export function allOf(...predicates: Predicate[]): Predicate {
  return (ctx) => predicates.every((predicate) => predicate(ctx));
}

/** A predicate that passes when any `predicates` member passes. */
export function anyOf(...predicates: Predicate[]): Predicate {
  return (ctx) => predicates.some((predicate) => predicate(ctx));
}

/** A predicate that passes when `predicate` fails. */
export function not(predicate: Predicate): Predicate {
  return (ctx) => !predicate(ctx);
}

File: sdks/java/src/main/java/org/byteveda/taskito/predicates/Predicates.java
package org.byteveda.taskito.predicates;

import java.util.List;

/** Combinators for building compound {@link Predicate}s. */
public final class Predicates {
    private Predicates() {}

    /** Passes only when every {@code predicate} passes (vacuously true when none). */
    public static Predicate allOf(Predicate... predicates) {
        List<Predicate> all = List.of(predicates);
        return context -> {
            for (Predicate predicate : all) {
                if (!predicate.test(context)) {
                    return false;
                }
            }
            return true;
        };
    }

    /** Passes when at least one {@code predicate} passes (false when none). */
    public static Predicate anyOf(Predicate... predicates) {
        List<Predicate> any = List.of(predicates);
        return context -> {
            for (Predicate predicate : any) {
                if (predicate.test(context)) {
                    return true;
                }
            }
            return false;
        };
    }

    /** Inverts {@code predicate}. */
    public static Predicate not(Predicate predicate) {
        return context -> !predicate.test(context);
    }
}
